Source: http://sourceware.org/ml/libc-alpha/2006-04/msg00090.html
Upstream: Ignored it apparently.
Reason: Compilation with gcc 4.5 and -march=i686 broke like this.

If gcc is configured to generate i686 code or better by default (like 
when using the --with-arch=pentium3 configure option), then the __i686
macro will always be defined automatically and thus screw up the
compilation of some .S files.
http://bugs.gentoo.org/131108

2006-04-25  Mike Frysinger  <vapier@gentoo.org>

	* sysdeps/i386/sysdep.h (__i686): Undefine.

--- sysdeps/i386/sysdep.h
+++ sysdeps/i386/sysdep.h
@@ -17,6 +18,14 @@
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA.  */
 
+/*
+ * When building for i686 targets or better, gcc automatically defines
+ * '__i686' to '1' for us which causes trouble when using section names
+ * like '__i686.get_pc_thunk.reg'.  Since we check for __i686__ in the
+ * code, killing '__i686' shouldn't be a problem.
+ */
+#undef __i686
+
 #include <sysdeps/generic/sysdep.h>
 
 #ifdef	__ASSEMBLER__

2006-04-25  Mike Frysinger  <vapier@gentoo.org>

	* sysdeps/pthread/pt-initfini.c: Include sysdep.h.

--- nptl/sysdeps/pthread/pt-initfini.c
+++ nptl/sysdeps/pthread/pt-initfini.c
@@ -1,5 +1,5 @@
 /* Special .init and .fini section support.  Linuxthread version.
-   Copyright (C) 1995,1996,1997,2000,2001,2002 Free Software Foundation, Inc.
+   Copyright (C) 1995-1997,2000-2002,2006 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
 
    The GNU C Library is free software; you can redistribute it
@@ -45,6 +45,9 @@
 /* Embed an #include to pull in the alignment and .end directives. */
 asm ("\n#include \"defs.h\"");
 
+/* Embed an #include to pull in asm settings. */
+asm ("\n#include <sysdep.h>");
+
 /* The initial common code ends here. */
 asm ("\n/*@HEADER_ENDS*/");
 
